h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java

```diff
@@ -283,79 +282,72 @@ void checkAndInheritModelProperties() {
+        throw new H2OIllegalArgumentException("Base model is not supervised: "+aModel._key.toString());
       }
 
       if (beenHere) {
         // check that the base models are all consistent with first based model
 
         if (modelCategory != aModel._output.getModelCategory())
-          throw new H2OIllegalArgumentException("Base models are inconsistent: there is a mix of different categories of models: "+Arrays.toString(_parms._base_models));
-
-        // NOTE: if we loosen this restriction and fold_column is set add a check below.
-        Frame aTrainingFrame = aModel._parms.train();
-        if (trainingFrameRows != aTrainingFrame.numRows() && !this._parms._is_cv_model)
-          throw new H2OIllegalArgumentException("Base models are inconsistent: they use different size (number of rows) training frames." +
-              " Found: "+trainingFrameRows+" (StackedEnsemble) and "+aTrainingFrame.numRows()+" (model "+k+").");
+          throw new H2OIllegalArgumentException("Base models are inconsistent: "
+                  +"there is a mix of different categories of models among "+Arrays.toString(_parms._base_models));
 
         if (! responseColumn.equals(aModel._parms._response_column))
-          throw new H2OIllegalArgumentException("Base models are inconsistent: they use different response columns." +
-              " Found: "+responseColumn+"(StackedEnsemble) and "+aModel._parms._response_column+" (model "+k+").");
-        
-//        if (blending_mode && _parms._blending.equals(aModel._parms._train)) {
-//          throw new H2OIllegalArgumentException("Base model `"+k+"` was trained with the StackedEnsemble blending frame.");
-//        }
-
-        // TODO: we currently require xval; loosen this iff we add a separate holdout dataset for the ensemble
-        if (cv_required_on_base_model) {
-          if (aModel._parms._fold_assignment != basemodel_fold_assignment) {
-            if ((aModel._parms._fold_assignment == AUTO && basemodel_fold_assignment == Random) ||
-                (aModel._parms._fold_assignment == Random && basemodel_fold_assignment == AUTO)) {
-              // A-ok
-            } else {
-              throw new H2OIllegalArgumentException("Base models are inconsistent: they use different fold_assignments.");
-            }
-          }
-
-          // If we have a fold_column make sure nfolds from base models are consistent
-          if (aModel._parms._fold_column == null && basemodel_nfolds != aModel._parms._nfolds)
-            throw new H2OIllegalArgumentException("Base models are inconsistent: they use different values for nfolds.");
+          throw new H2OIllegalArgumentException("Base models are inconsistent: they use different response columns."
+                  +" Found: "+responseColumn+"(StackedEnsemble) and "+aModel._parms._response_column+" (model "+k+").");
+
+        if (require_consistent_training_frames) {
```

Comment:
@seb-h2o I believe there is still a chance that training frames are different even though number of rows is not. We decided not to consider this case as it is a highly unlikely use case?

Reply:
@deil87 I don't think it's a question of likely/unlikely.
It is just a question of what is necessary for StackedEnsemble to be able to stack the models.

When using xval, we allow the various models to be trained with different training frames, they may have a different set of predictors or have different feature engineering for example, but the response column should be the same, and xval predictions MUST have the same number of rows, that's all. If user is using a completely different training frame with the same number of rows, and a response column with a similar name (e.g. "class"), then we won't detect it. Maybe we could store the hash of the response column for additional safety? Wasn't done until now.

When using blending, the rows restriction is even removed, remains only the name/type of the response... you could almost blend orange with apples, which is not that uncommon in juice.